File: typescript/src/safety-number.ts
import { autograph_init, autograph_safety_number } from './clib'
import { createSafetyNumberBytes } from './utils'
import { InitializationError, SafetyNumberCalculationError } from './error'

const calculateSafetyNumber = async (
  a: Uint8Array,
  b: Uint8Array
): Promise<Uint8Array> => {
  if ((await autograph_init()) < 0) {
    throw new InitializationError()
  }
  const safetyNumber = createSafetyNumberBytes()
  const success = autograph_safety_number(safetyNumber, a, b) === 0
  if (!success) {
    throw new SafetyNumberCalculationError()
  }
  return safetyNumber
}

export default calculateSafetyNumber

File: rust/src/safety_number.rs
use alloc::vec::Vec;

use crate::clib::{autograph_init, autograph_safety_number};
use crate::error::Error;
use crate::utils::create_safety_number_bytes;

pub fn calculate_safety_number(a: &Vec<u8>, b: &Vec<u8>) -> Result<Vec<u8>, Error> {
    if unsafe { autograph_init() } < 0 {
        return Err(Error::Initialization);
    }
    let mut safety_number = create_safety_number_bytes();
    let success =
        unsafe { autograph_safety_number(safety_number.as_mut_ptr(), a.as_ptr(), b.as_ptr()) } == 0;
    if success {
        Ok(safety_number)
    } else {
        Err(Error::SafetyNumberCalculation)
    }
}
